decisions: improving process

markus2330 opened the pull request:
<!--
Check relevant points but **please do not remove entries**.
-->

## Basics

<!--
These points need to be fulfilled for every PR.
-->

- [ ] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [ ] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - [ ] add a line in `doc/news/_preparation_next_release.md`
  - [ ] reformat the code with `scripts/dev/reformat-all`
  - [ ] make all unit tests pass
  - [ ] fix all memleaks
- [ ] The PR is rebased with current master.

<!--
If you have any troubles fulfilling these criteria, please write
about the trouble as comment in the PR. We will help you,
but we cannot accept PRs that do not fulfill the basics.
-->

## Checklist

<!--
For docu fixes, spell checking, and similar none of these points below
need to be checked.
-->

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation (see [Documentation Guidelines](https://www.libelektra.org/devgettingstarted/documentation))
- [ ] I fixed all affected decisions (see [Decision Process](https://www.libelektra.org/decisions/decision-process))
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://www.libelektra.org/devgettingstarted/coding))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned [every code](/.reuse/dep5) not directly written by me in [reuse syntax](https://reuse.software/)

## Review

<!--
Reviewers should check the following.
-->

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

<!--
If you are already Elektra developer, please adjust the labels.
Otherwise, write a comment and it will be done for you.
-->

- [ ] Add the "work in progress" label if you do not want the PR to be reviewed yet.
- [ ] Add the "ready to merge" label **if the basics are fulfilled** and no further pushes are planned by you.

markus2330 commented:
I changed that "In Discussion" is not mandatory anymore. Instead "Decided" is now mandatory, which is now also the only step which must be in form of a decision PR. I.e. decisions of other steps don't necessarily need the formal "decision PR" review with only one decision and two reviewers.

I hope this makes the process light-weight enough and addresses the concerns. I also added an assumption that the decision process by itself should **not** be a obstacle other than writing more documentation up-front if someone considers this as obstacle.

kodebach commented on doc/decisions/STEPS.md:
```diff
@@ -0,0 +1,86 @@
+
+## Drafts
+
+> This step is recommended if the problem is not yet clear.
```
"Clear" to whom? In most cases the problem will be clear to the decision author.

Right now it's not at all clear when this step should be taken. With the extra sentence I suggested above, it would be clearer IMO. A decision always starts a "Draft". It can be merged as a "Draft", but it can also progress to "In Discussion" or even "Decided" before the merge.

kodebach commented on doc/decisions/STEPS.md:
````diff
@@ -0,0 +1,86 @@
+    In_Progress --o Delayed
+```
+
+## Drafts
````
```suggestion
The first PR for a decision must **always** create the decision in the "Draft" state at first.
If during the reviews of this PR it becomes clear that the decision is further along, can be moved to "In Discussion" or "Decided" before the merge.

## Drafts
```

IMO this is a better process, because it makes clear how the first PR should be created. Otherwise the decision author needs to guess what the correct initial state of the decision is.

kodebach commented on doc/decisions/STEPS.md:
```diff
@@ -0,0 +1,86 @@
+Here the decision should not only have one decision but should describe several solutions.
+For each solution a proposal, rationale and optionally implications should be given.
+
+## In Progress
```
How is this different from "In Discussion" when would we use this?

IMO this step should be removed.

kodebach commented on doc/decisions/STEPS.md:
```diff
@@ -0,0 +1,86 @@
+> Only exceptions like backlinks from other decisions are allowed.
+
+## Partially Implemented
+
```
```suggestion

> This step is optionally used when a decision cannot be implemented in a single PR.

```

kodebach reviewed (changes requested):
LGTM for the most part. Some minor comments and one bigger issue: I think its not clear what it means if a step is mandatory or recommended.

AFAICT "mandatory" means we must merge a PR that puts the decision into this state. So there has to be at least one PR that puts the decision into "Decided" and another one that puts it into "Implemented".

I think "recommended" should be replaced with the term "optional". Recommended suggests we should try to hit that state. That would mean that e.g. if possible the decision should be merged once in the "In Discussion" state. IMO that would be nonsense. Merging in that state can make sense, if the PR gets too big. But the merge is totally optional.

Also I think decisions should be initially created in the "Draft" state. That may already be implied, but it should be explicitly stated.

markus2330 commented:
@kodebach wrote in https://github.com/ElektraInitiative/libelektra/pull/4641#issuecomment-1311530661
> I need to decide myself, what to propose first.

Please don't fix your mindset on a particular solution, this is a big call for trouble. As written in the decision process, drafts should focus on the problem, so no "best solution" should be proposed in the beginning. I clarified in 7d63f4d36d5e21307d0158f157d37f820524b833

If you simply meant that you need to think about the problem, then go on thinking :revolving_hearts:

kodebach commented:
> Please don't fix your mindset on a particular solution, this is a big call for trouble.

I'm not doing that. Although I will have thought about a solution (or probably even multiple solutions), simply because that's how I understand a problem. 

I even stated the actual reason in my original comment:

> I have multiple ideas right now and don't want to with another 10 decision PRs that are all stalled

There is no point in creating 10 decision PRs at the same time. Unless they are closely related, discussing multiple problems at the same time just becomes confusing and means people are less focused.

> I clarified in https://github.com/ElektraInitiative/libelektra/commit/7d63f4d36d5e21307d0158f157d37f820524b833

You really don't have to clarify every little detail. The description of the decision process is already very long. At some point adding more details will just make it more confusing for newcomers. If we really want to record all the details, maybe we need two files.

1. A file that gives an overview of the decision process for newcomers (e.g. `doc/decisions/README.md`)
2. The decision file about the decision process, which holds all the details and extra clarifications. This file should only be consulted, if something is unclear in the first file or there is a disagreement on the interpretation of the first file.

markus2330 commented:
> You really don't have to clarify every little detail.

If you look at our discussions in the past: it is not an detail but it is essential. We often had long discussions where we went back and forth from solution to the problem. E.g. our discussions about the basename. Decisions where we had the problem right from the beginning are a much smoother experience for everyone! :heart: 

I don't think there is a useful way to abbreviate the description about the process for newcomers without losing essentials.

kodebach commented:
> If you look at our discussions in the past

That is exactly my point. _Sometimes_ these specific clarifications are needed. But they should be in a separate place. If the general description of the process explains the solution for every little edge case, it can become very hard to understand the what the core ideas are.

> I don't think there is a useful way to abbreviate the description

I have not tried to summarize it myself. Maybe there isn't anything to cut right now. All I wanted to say is that we should not infinitely add more clarifications. At some point it becomes overwhelming and too detailed.

kodebach commented on doc/decisions/STEPS.md:
```diff
@@ -0,0 +1,98 @@
+- Now it is allowed to have the decision from the previous round in the "Decision" section.
+
+> For decisions "In Progress", unlike "In Discussion", also the solution space next to the problem is clear.
+> It is the highest of the steps reachable for a decision PR that started in "Drafts".
```
IMO that sentence is pretty confusing.

For one, I'd say "Decided" is the highest reachable step. Because I could create a PR with the decision as "Drafts" and move it trough the stages. Only when it reaches "Decided" do we require a merge, before we can move on.

Also the bit about "started in Drafts" is confusing. If I start the decision in "In Discussion" I could move beyond this step, or what?

markus2330 replied:
There is also the rule: "If during the reviews of this PR it becomes clear that the decision is further along, can be moved to "In Discussion" or "In Progress" before the merge." (which is the full list), so it is not possible to directly go to Decided from Drafts. See also updated Diagram: 674ec185e894c13cc22edb5d547352ef6ef70e8a

I changed the confusing sentence to: "Without merges in between, this is the last step reachable for a decision PR that started in "Drafts"." 570e48ab307995a791d7340d62cbbd23139e1331

kodebach commented on doc/decisions/STEPS.md:
```diff
@@ -0,0 +1,98 @@
+- You must include all further alternative proposals made in the "Considered Alternatives" section.
+- Now it is allowed to have the decision from the previous round in the "Decision" section.
+
+> For decisions "In Progress", unlike "In Discussion", also the solution space next to the problem is clear.
```
This doesn't help me one bit. I still don't understand the when "In Discussion" becomes "In Progress", but not "Decided" yet.

To me this feels like a valid step in the theoretical process, but doesn't fit with using PRs. IMO it sounds like a decision that has a decision (in the "Decision" section), but is not merged yet (because reviews aren't done) would be "In Progress". But that is a useless distinction. That only means you need to move the decision file, after the reviews are done.

When would we merge a decision as "In Progress"?

If I look at the current decision in [`doc/decisions/2_in_progress`](https://github.com/ElektraInitiative/libelektra/tree/master/doc/decisions/2_in_progress), the picture doesn't become much clearer. Many of them are "Decided" AFAIAC, some are even "Partially Implemented". Only `simplify_api.md` is still not decided IMO, but I'd say that's "In Discussion" because it is not clear what the exact API changes will be.

markus2330 replied:
> This doesn't help me one bit.

I elaborated the explanation in 5a330e8e85f47200d8dd965579f696855df00eac

> I still don't understand the when "In Discussion" becomes "In Progress", but not "Decided" yet.

Should be answered above. I feel like a merge point before we finally decide is cleaner.

> That only means you need to move the decision file, after the reviews are done.

For that we need good solutions anyway.

> When would we merge a decision as "In Progress"?

Always. It is a must, as it cannot be directly decided. (Unless it was directly target as decided for decisions where no discussion is needed. E.g. a decision needs another decision, split up of decisions etc..)

> Many of them are "Decided" AFAIAC, some are even "Partially Implemented".

These are left-overs before we had the decision process. I started cleanup in  #4690. The cleanup won't be finished with  #4690, though.

kodebach replied:
I think I get the "big picture" now see https://github.com/ElektraInitiative/libelektra/pull/4639#discussion_r1027069545. What I still don't quite understand is: Why we need both "In Discussion" and "In Progress"? Is it just a bookkeeping thing to know how close to a decision we are?

Maybe different names would be better (open to other suggestions):
- "Drafts" -> "Problem Draft" (because the problem is still a draft)
- "In Discussion" -> "Solution Draft" (problem is done, but considered alternatives, i.e. solutions, are still drafts)
- "In Progress" -> "In Review" (because theoretically everything is done, the review process just hasn't come to a decision yet)

If we want to keep the current names, IMO "In Discussion" and "In Progress" should be swapped. First it is "In Progress", i.e. we are working on possible solutions, _then_ it is "In Discussion", where we discuss only which solution is best.

markus2330 replied:
This is excellent input. I fully agree not only that the names are confusingly worded but trying to find better names also also something else became clear to me: Drafts should not have **any** requirements to be accepted. E.g. #4683 clearly is a well-written draft (clear problem, elaborate considered alternatives etc.). Just because of our definition of Drafts (that reviewers must agree on the problem), the PR is blocked. This hampers the possibility to collect proposals for problems that might occur in future. If we would already have had this draft in our drawer before thinking of COW, we would have probably implemented COW differently. But as COW is done now, the previous interesting decision now doesn't have a clear problem anymore. I agree that "Rejecting" such a decision is the wrong word. So I think it should be allowed to be merged as draft. If later another decision would benefit from having this decision, we can easily refer to it again.

I updated the decision process in 7691244f94ce63b266396546edfae56028bc3099
@kodebach Please review.

kodebach commented on doc/decisions/STEPS.md:
````diff
@@ -0,0 +1,105 @@
+
+This document describes all steps a decision can run through.
+
+```mermaid
+flowchart LR
+  s((Start)) --> Drafts --> In_Discussion --> In_Progress -- merge --> Decided
+    -- merge --> Partially_Implemented --> Implemented
+
+  %% Shortcuts:
+  s --> Decided
+  Drafts -- merge --> Decided
+  In_Discussion -- merge --> Decided
+  Decided -- merge --> Implemented
+```
````
This needs a bit more explanation. What does the "merge" on the edge mean?

I guess, what you want to say is:

1. The first PR for a decision can be created as "Drafts", "In Discussion", "In Progress" or "Decided"
2. A decision PR that wants to merge a decision as "Decided" must:
    1. Already start with the decision as "Decided"
    2. Already have all the considered alternatives fully explored
    3. Only contain discussions/reviews about deciding between one of the considered alternatives
    4. If there is any significant change in the PR (i.e. information changes, not just typos) then the PR cannot be merged as "Decided". The decision must be moved back to "In Progress"
3. After a decision is merged as "Decided", it can be moved to "Partially Implemented" or "Implemented" by any PR.

If that is correct, maybe add this text or something similar in addition to the diagram.

markus2330 replied:
Thanks for the summary, I added it in bbe44da7c9d821ece2aecc2a1c400061302cf9d6

Furthermore I wrote:
> The label `merge` on the edges mean that a decision PR must be merged before it can target the next step.

markus2330 commented:
@lukashartl Is the little problem you talked about still present?

lukashartl commented on doc/decisions/5_implemented/decision_process.md:
```diff
@@ -73,21 +75,20 @@ Decisions by supervisors are undemocratic and decisions in meetings are nontrans
   Nevertheless they don't want to avoid other solutions, are open to arguments/facts/etc. and incorporate all input of decision PR fairly.
 - We will always be able to reach an consensus even if it requires that the core or plugins get multiple implementations.
   We don't need a vote (besides the approved review) or a benevolent dictatorship.
```
Do you really think that there will always be a consensus?
Is having multiple implementations in the core or in plugins really a good idea? What if future decisions lead to adoptions that wouldn't be possible in all implementations? Wouldn't that add technical debt in the long run?

markus2330 replied:
> Do you really think that there will always be a consensus?

I updated:

```
+- For important decisions, we will able to reach an consensus even if it requires that the core or plugins get multiple implementations.
+  We accept indefinite postponement of lesser important decisions, where there is strong disagreement about the problem or the best solution.
+  Thus we don't need a vote (besides the approved review) or a benevolent dictatorship.
```

0a7d274b9209d12b37856dca40067c238f3ad253

> Is having multiple implementations in the core or in plugins really a good idea?

If there are different requirements/goals: Why not?

> What if future decisions lead to adoptions that wouldn't be possible in all implementations?

I don't know, do you have an example of what this could be?

lukashartl commented:
> @lukashartl Is the little problem you talked about still present?

I've added my feedback (review) to the corresponding line.
Great to see the the mermaid.js diagram :+1:

kodebach commented on doc/decisions/STEPS.md:
```diff
@@ -54,28 +78,34 @@ Here you must ensure:
 - "Decision", "Rationale" and "Implications" are now filled out and fixed according to the reviews
 - decisions of this step usually already have an implementation PR
 
+Decisions that need an update, e.g. because assumptions changed, usually directly start with the step "Decided".
```
I think that needs to be phrase differently. IMO it should only start in "Decided", if the chosen solution stays the same and only details change. If the problem it self changes, or the chosen solution is no longer a good choice, then we should probably start in an earlier step.

markus2330 replied:
Changed usually -> sometimes. I don't think we should put rules on rare cases. e447fde4d313e45aafc416f7073a520845dc8dba

kodebach commented on doc/decisions/STEPS.md:
```diff
@@ -90,9 +120,20 @@ The "Implication" must clearly say how much of the decision is already implement
+
+This step is:
+
+- for decisions that are useful contributions but currently there is simply no urgency or immanent need for the proposal.
```
```suggestion
- for decisions that would be useful contributions but there is currently nobody with the time to focus on the problem.
```

If there is a clear problem and there is a person who wants to work on it, why not let them? They may not want to or have the skill to work on other problems, so refusing their contribution would just be a waste.

kodebach commented on doc/decisions/STEPS.md:
```diff
@@ -90,9 +120,20 @@ The "Implication" must clearly say how much of the decision is already implement
-Alternatively, decisions might be rejected (i.e. status quo wins).
-These decision PRs are also merged for documentation purposes.
+Alternatively, decisions might be rejected (e.g. status quo wins).
+This step is for decisions that are agreed to not fit well to Elektra and never should be accepted.
```
```suggestion
This step is for decisions for which a consensus exists that the decision should not be taken right now.
If circumstances change, the decision may be revisited.
```

kodebach reviewed (changes requested):
In general I like the latest changes. However, I dislike the names "Alternatives Clear" and "Problem Clear". Not only is the word "clear" vague, the name also doesn't describe the current status of the decision. It describes the previous history.

IMO there is also still one step too many:
- "Drafts": We define the problem.
- "Problem Clear": Problem now fixed, we define the solutions (=alternatives).
- "Alternatives Clear": Alternatives now fixed, we do what exactly?
- "In Review": Everything fixed, we just talk about pros/cons and evaluate our options.
- "Decided": Everything fixed, and we have chosen a solution.

I would use these steps:
- "Draft": We define the problem.
- "Exploration": We explore solutions for the problem.
- "Review": We review the different options, try to choose one.
- "Decided": We have chosen a solution.

Also, if we're already talking about terms. I really dislike the way "Considered Alternatives" is used as a term. It makes sense as a title for the section in the template. But as a term I would use "options" or "possible solutions" or maybe "considered solutions". The term "alternatives" only really makes sense in conversation, once a decision has been made. The term "solution" also makes more sense IMO, because we have a problem and are looking for a solution.

markus2330 commented on doc/decisions/STEPS.md:
````diff
@@ -4,17 +4,36 @@ This document describes all steps a decision can run through.
 ```
 
-Additionally, decision that are not yet "Decided" can be become "Rejected" or "Delayed" at any point.
+Additionally, decision that are not yet "Decided" can be become "Drafts", "Rejected" or "Postponed" at any point, e.g., if the decision author stops working on the decision.
````
I changed can -> may. This is something rare, so I don't think we need to write rules for this. 00ddc456120d01b145a473e31904966875886ead

markus2330 commented:
> However, I dislike the names "Alternatives Clear" and "Problem Clear". Not only is the word "clear" vague, the name also doesn't describe the current status of the decision. It describes the previous history.

Then we need to make the word "clear" more clear.

I wrote in 8a20654330b68c98183b58c3a65d926d765b86dd:
> A problem is clear if everyone would be able to describe an experiment or test case that shows if a solution fixes the problem.

> The alternatives are clear if all reviewers understand the given alternatives and no reviewers can come up with new alternatives.

You wrote:
> IMO there is also still one step too many:

>   "Problem Clear": Problem now fixed, we define the solutions (=alternatives).
>   "Alternatives Clear": Alternatives now fixed, we do what exactly?

If someone said "I have one more idea", it cannot be merged in "Alternatives Clear" but it can be merged in "Problem Clear".

> I would use these steps:

>    "Draft": We define the problem.
>    "Exploration": We explore solutions for the problem.
>    "Review": We review the different options, try to choose one.
>    "Decided": We have chosen a solution.

There is no problem if a step stays empty but there is a big problem if there is uncertainty about the step a decision is currently in. In your suggestion it might be unclear if the reviewers came to an agreement about the problems or solutions.

> Also, if we're already talking about terms. I really dislike the way "Considered Alternatives" is used as a term. It makes sense as a title for the section in the template. But as a term I would use "options" or "possible solutions" or maybe "considered solutions".

I don't understand what you mean here. The title for the section must be the chosen term.

> The term "alternatives" only really makes sense in conversation, once a decision has been made. The term "solution" also makes more sense IMO, because we have a problem and are looking for a solution.

"Considered alternatives" makes more sense if it only contains solutions that were **not** taken, hence "alternatives". In a few recent PRs we (also @atmaxinger) misused it a bit and referred from the decision to one of the considered alternatives. (which makes sense because of how our process works.)  If we want to keep this style, we probably should rename the section. "Solutions" makes sense to me.

kodebach commented on doc/decisions/STEPS.md:
```diff
@@ -0,0 +1,141 @@
+Here the decision should not only have one decision but should describe several solutions.
+For each solution a proposal, rationale and optionally implications should be given.
+
+> The alternatives are clear if all reviewers understand the given alternatives and no reviewers can come up with new alternatives.
```
"no reviewers can come up with new alternatives" is a very bad criteria IMO. You can (almost) always come up with more alternatives. It also necessarily leads to long lists of alternatives that were never seriously considered because they are obviously flawed. But somebody had the idea, so we have to list them according to this definition...

markus2330 replied:
I agree, f832f42104bef6ada4da94c061cf698d928a19ff

```diff
-> and no reviewers can come up with new alternatives.
+> and no reviewer can come up with better alternatives.
```

kodebach commented:
> > doesn't describe the current status of the decision. It describes the previous history.
> 
> Then we need to make the word "clear" more clear.

The part about not describing the current status, was actually far more of a concern. I don't like the terms "X clear" and I don't think additional definitions will change that. I can accept that I'm in the minority here, but I still think the way the steps are described right now is a bit confusing. Not because of missing definitions, but because from the name of the "step" it's not obvious whats happening with the decision right now.

The other terms also describe what _has happened_ not what _is happening_ or _will happen_, but IMO in these cases it is more obvious what the next step is:

- "Decided": We decided to do something, we just haven't done it yet. I.e. we know what we will do, we just have to start doing it now.
- "Partially Implemented": Some of it is done, some is still to do.
- "Implemented": It's done, nothing to do anymore.
- "Rejected": It's rejected, so clearly nothing will happen anymore.
- "Delayed": It's delayed. Seems obvious that nothing will happen right now, but maybe at some point in the future it will be revisted.

But with the first 3 steps, it's not that clear whats going on and what's going to happen. How about these alternatives:

- "Draft" -> "Problem Draft": Makes it clear the problem itself is still a draft.
- "Problem Clear" -> "Solution Draft": Makes it clear the solutions are still drafts and unfinished.
- "Alternatives Clear" -> "In Review": Merge the two steps. I don't understand the difference. In this step the problem and the solutions are finished, we're just gathering opinions and data to make a decision.

> If someone said "I have one more idea", it cannot be merged in "Alternatives Clear" but it can be merged in "Problem Clear".

Okay, I get that, but then what is the distinction between "Alternatives Clear" and "In Review"? 

> I don't understand what you mean here. The title for the section must be the chosen term.

What I dislike is that "Considered Alternative" is used as a single term. If we used the term "alternatives" for the different options/solutions/suggestions then that would also be fine (although I still prefer "solution" because that fits better with "problem").

The title of the section can still be "Considered Alternatives" or "Considered Solutions". Because these are the alternatives/solutions we considered. But one of the subsection should not be "a considered alternative"/"a considered solution". It should just be "a alternative"/"a solution" IMO. (The "considered" is implied by the fact we're talking about it anyway).

> In a few recent PRs we (also atmaxinger) misused it a bit and referred from the decision to one of the considered alternatives.

I actually like this approach because it makes the PRs smaller (no big diff that just moves the text down to the "Decision" section) and IMO makes the file easier to read after the fact. By reading from top to bottom you learn the problem ("Problem", "Constraints", "Assumptions"), the solutions we explored ("Considered Solutions") and then what we decided ("Decision"), why ("Rationale") and what that means ("Implications").

If we strictly go with alternatives the flow is more: Whats the problem? What didn't we do? What did we do? Why and what does it mean?. Seems a bit odd to discuss the things we didn't do before we describe the decision. So if you want to go strictly with "Alternatives", then I'd actually move "Considered Alternatives" to the bottom of the file, to make the reading flow better.

markus2330 commented:
> The part about not describing the current status, was actually far more of a concern. I don't like the terms "X clear" and I don't think additional definitions will change that. I can accept that I'm in the minority here, but I still think the way the steps are described right now is a bit confusing. Not because of missing definitions, but because from the name of the "step" it's not obvious whats happening with the decision right now.

At some point terms stop working by itself and definitions are needed.

> what is the distinction between "Alternatives Clear" and "In Review"?

Thank you, there was an error in the description which seemingly confused you. I fixed it in 0aa11cd830a8b12e191c51937174935f434504d1

To also answer the question: That the decision is now in the "Decision" section. In "In Review" mainly the one decision is reviewed.

>     I don't understand what you mean here. The title for the section must be the chosen term.
> 
> What I dislike is that "Considered Alternative" is used as a single term. If we used the term "alternatives" for the different options/solutions/suggestions then that would also be fine (although I still prefer "solution" because that fits better with "problem").

In http://eprints.cs.univie.ac.at/2345/1/02_Using_Patterns_to_Capture.pdf "Positions" is chosen. "Solutions" is also fine, I did the change in 9b91ed87d72802b62bcda4721e7f482de1213198

> >     In a few recent PRs we (also atmaxinger) misused it a bit and referred from the decision to one of the considered alternatives.
> 
> I actually like this approach

I explicitly allowed it now in the explanation.

> because it makes the PRs smaller (no big diff that just moves the text down to the "Decision" section) and IMO makes the file easier to read after the fact. By reading from top to bottom you learn the problem ("Problem", "Constraints", "Assumptions"), the solutions we explored ("Considered Solutions") and then what we decided ("Decision"), why ("Rationale") and what that means ("Implications").

This order you have anyway but I agree that "Decision" in a more summary style referring to the solutions above is a nice quick read, especially for decisions where the solutions are very long.

kodebach commented:
> At some point terms stop working by itself and definitions are needed.

True, but "solutions" is a term that can be understood correctly for the context, just by knowing the English language definition. "Considered Alternatives" is much more vague, if you don't know our specific definition of it. That's what I meant. Also using "considered alternatives" as a term can lead to some odd sounding grammatical constructions, while "solutions" is just a normal word.



> "Solutions" is also fine, I did the change in [9b91ed8](https://github.com/ElektraInitiative/libelektra/commit/9b91ed87d72802b62bcda4721e7f482de1213198)

There are still some places, where "alternatives" is used and the existing decision files still use "Considered Alternatives". No need to fix immediately, but we shouldn't forget.

What should be fixed right now, is that "Alternatives Clear" as a step name makes little sense, if the term is now "solutions". It should be "Solutions Clear".

> Thank you, there was an error in the description which seemingly confused you. I fixed it in [0aa11cd](https://github.com/ElektraInitiative/libelektra/commit/0aa11cd830a8b12e191c51937174935f434504d1)


Okay, now I understand the distinction, but I still don't know _why_. What is the point of merging a decision as "In Review" instead of keeping it in "Solutions Clear", until it can be moved to "Decided"?

Also I think it could be _very confusing_ to have decision in the repo with content in "Decision", which is not yet "Decided".

markus2330 commented:
There was a merge conflict, I rebased. Maybe it can already accepted as it is. Renaming in decisions itself should happen in a later PR.

> What should be fixed right now, is that "Alternatives Clear" as a step name makes little sense, if the term is now "solutions". It should be "Solutions Clear".

Thanks, fixed in d7be428c47fc9b2dbdecf1e281bf69272395830f

>  Okay, now I understand the distinction, but I still don't know why. What is the point of merging a decision as "In Review" instead of keeping it in "Solutions Clear", until it can be moved to "Decided"?

"In Review" and "Decided" decisions have the same content. It is about that "Decided" should be an extra step, so that they cannot slip through from a draft where the reviewers maybe didn't see that it already targeted "Decided", or a bigger PR which included the decision. If it slips through until "In Review", no harm is done.

In c4e09a26d1b6f43b6a24b2c5c8288efabe61a4da I wrote: "As "Decided"/"Implemented" decisions need a special decision PR, we cannot merge a "Decided"/"Implemented" decision PR by accident."

> Also I think it could be very confusing to have decision in the repo with content in "Decision", which is not yet "Decided".

As decisions are in extra folders according to their step, I think we'll get used to it quite quickly.

markus2330 commented:
Second reviewer, please feel free to merge.

kodebach commented on doc/decisions/EXPLANATIONS.md:
```diff
@@ -96,6 +110,6 @@ If particular information is important and not present in any sections above, pl
 
 Any incomplete and unexplored idea/opinion, which is not complete enough to be in "Considered Alternatives", can be written here.
```
```suggestion
Any incomplete and unexplored idea/opinion, which is not complete enough to be in "Solutions", can be written here.
```

kodebach commented on doc/decisions/STEPS.md:
```diff
@@ -0,0 +1,144 @@
+- problem, constraint and assumptions are well-explained and sound
+- links from/to related decisions are created
+- there are several solutions described, each with rationale and implication
+- "Decision", "Rationale" and "Implications" are **not** yet filled out if there are people arguing for different options
```
```suggestion
- "Decision", "Rationale" and "Implications" are **not** yet filled out
```

If people don't argue for different options, this would be completely the same as "In Review". So better just say "not filled out at all" here.

kodebach commented on doc/decisions/STEPS.md:
```diff
@@ -0,0 +1,144 @@
+> I.e. the trade-offs, combinations and pros/cons of the solutions are explored.
+> Decision author and reviewers are satisfied that every useful solution is present in the decision.
+
+## In Review
+
+Now it is allowed to have the decision from the previous round in the "Decision" section.
+In this step, the last details of the chosen decision get polished:
+
+- consistency with other decisions
+- links from/to related decisions checked
+- "Decision", "Rationale" and "Implications" are fully filled out
```
IMO this is just describes the state of an **open/unmerged** PR that wants to move/add a decision file to `doc/decisions/X_decided`. I don't really see a reason to ever merge a file in this state, _without_ also having the reviews and committing to the decision.

To me merging a fully filled out decision file in `doc/decisions/Y_in_review` seems like a way of creating a bias towards the decision the author wants. If the reviews already agree, then there is no need for this step. If reviews disagree merging the "Decision" section anyway seems kinda wrong.

markus2330 replied:
> merge a file in this state, without also having the reviews and committing to the decision.

There might be plenty of reasons, like one of the reviewers is not available anymore.

> seems like a way of creating a bias towards the decision the author wants

Yes but in this case it is allowed because there were already (hopefully unbiased) discussions showing this clear tendency.

kodebach commented on doc/decisions/STEPS.md:
```diff
@@ -0,0 +1,144 @@
+- links from/to related decisions checked
+- "Decision", "Rationale" and "Implications" are fully filled out
+
+> Without merges in between, this is the last step reachable for a decision PR that started in "Drafts".
```
IMO creating a PR that introduces a new decision file and adds it directly to `doc/decisions/X_decided` should be fine. Since we need the reviews anyway, such a PR only makes sense, if the decision is only documenting status quo, consequences of other decisions, etc. Basically only decisions were there is only one realistic option.

markus2330 replied:
If the decisions are only documentation is does not really matter.

kodebach reviewed (changes requested):
> "In Review" and "Decided" decisions have the same content.

Yes, I that's exactly why I think it's confusing and doesn't serve much purpose.

IMO the PR that puts a decision into "Decided" should also contain the commits that add the "Decision", "Rationale" and "Implications" sections. If there is disagreement about the contents of these sections, that should be worked out in this one PR. If the discussions become too much, then I'd say just merge it again as "Solutions Clear" and start a new attempt to move to "Decided" in a new PR.

> Second reviewer, please feel free to merge.

I think we still need more changes, so please don't merge yet.

Also, I **strongly** disagree with moving the `decision_process` file to `5_implemented`:
1. "Implemented" (to me) would mean that all the other decisions are up-to-date for this scheme, i.e. use "Solutions" section, we have all the steps, etc.
2. More importantly, I wouldn't even call this decision "Decided" anymore. The files referenced in the "Decision" section of `decision_process`, which describe core parts of the actual decision, are still changing significantly.
3. Just going by the steps described by you in this PR, I'd say `decision_process` should be "In Review". In my mind, it is at best in "Problem Clear", because clearly one of the solutions (the one we've decided on) is still changing significantly.

But that's more of a technicality. I'd just keep the file in `3_decided` (where it is now) and move it to `5_implemented` in the PR that updates the other decisions to the new terms/steps.

markus2330 commented:
> >   "In Review" and "Decided" decisions have the same content.

> Yes, I that's exactly why I think it's confusing and doesn't serve much purpose.

Its nevertheless a huge difference in the step, hence a separate step is obligatory.

> Also, I strongly disagree with moving the decision_process file to 5_implemented:

I agree, the current decision is not fully implemented anymore, I moved to partially_implemented.

kodebach commented:
> > > "In Review" and "Decided" decisions have the same content.
> 
> > Yes, I that's exactly why I think it's confusing and doesn't serve much purpose.
> 
> Its nevertheless a huge difference in what the step communicates, hence a separate step is obligatory.

I still don't see the point of having these separate steps, but I've given up... I guess it won't hurt to have the step, we'll see whether it gets used or remains purely theoretical anyway.

markus2330 commented:
Yes, if it is useless, we can easily remove the empty folder.

Everything rebased, please review and merge.

markus2330 commented:
jenkins build libelektra please

markus2330 commented:
jenkins build libelektra please

markus2330 commented:
@flo91@tmakar @mpranj @atmaxinger can someone please review&merge this?

atmaxinger reviewed (approved):
LGTM, describes the decision process very clearly. I also like the change in wording from "considered alternatives" to "solutions".
